compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp

```diff
@@ -459,13 +457,184 @@ struct DistributeBroadcast final : OpDistributionPattern<vector::BroadcastOp> {
+    }
+
+    // Flatten the locally reduced value.
+    VectorType shaped = locallyReduced.getType();
+    int64_t numElements = shaped.getNumElements();
+    SmallVector<int64_t> flatShape(1, numElements);
+    Type elemTy = shaped.getElementType();
+    VectorType flatVecType = VectorType::get(flatShape, elemTy);
+    VectorValue flat =
+        rewriter.create<vector::ShapeCastOp>(loc, flatVecType, locallyReduced);
+
+    FailureOr<VectorValue> globalReduced = doThreadGlobalReduction(
+        rewriter, srcLayout, flat, multiReduceOp.getKind(), reducedDims);
+    if (failed(globalReduced)) {
+      return failure();
+    }
+
+    VectorValue unflattened = rewriter.create<vector::ShapeCastOp>(
+        loc, shaped, globalReduced.value());
+    replaceOpWithDistributedValues(rewriter, multiReduceOp, unflattened);
+
+    return failure();
+  }
+
+  FailureOr<VectorValue>
+  doThreadGlobalReduction(RewriterBase &rewriter, NestedLayoutAttr layout,
+                          VectorValue flat, vector::CombiningKind kind,
+                          ArrayRef<bool> reductionMask) const {
+    VectorType flatVecType = flat.getType();
+    Type elemTy = flatVecType.getElementType();
+    int64_t numElements = flatVecType.getNumElements();
+    Location loc = flat.getLoc();
+
+    unsigned elemBitwidth = elemTy.getIntOrFloatBitWidth();
+    if (elemBitwidth != maxBitsPerShuffle) {
+      return rewriter.notifyMatchFailure(
+          flat.getDefiningOp(), llvm::formatv("unimplemented: packed shuffle",
+                                              elemBitwidth, maxBitsPerShuffle));
+    }
+
+    auto constOp = rewriter.create<arith::ConstantOp>(
+        loc, rewriter.getZeroAttr(flatVecType));
+    auto res = llvm::cast<VectorValue>(constOp.getResult());
+
+    for (unsigned i = 0; i < numElements; ++i) {
+      Value extracted = rewriter.create<vector::ExtractOp>(loc, flat, i);
+
+      // Reduce across all reduction dimensions 1-by-1.
+      for (unsigned i = 0; i < reductionMask.size(); ++i) {
+        if (reductionMask[i]) {
+          extracted = doPackedThreadReductionOnDim(rewriter, layout, extracted,
+                                                   kind, i);
+        }
+      }
+
+      res = rewriter.create<vector::InsertOp>(loc, extracted, res, i);
+    }
+
+    return res;
+  }
+
+  Value doPackedThreadReductionOnDim(RewriterBase &rewriter,
+                                     NestedLayoutAttr layout, Value val,
+                                     vector::CombiningKind kind,
+                                     int64_t dim) const {
+    Location loc = val.getLoc();
+    int64_t offset = getShuffleOffset(layout, dim);
+    int64_t width = getShuffleWidth(layout, dim);
+
+    // TODO: Query from entrypoint.
+    int64_t subgroupSize = 64;
+
+    for (int i = offset; i < offset * width; i <<= 1) {
+      auto shuffleOp = rewriter.create<gpu::ShuffleOp>(
+          loc, val, i, subgroupSize, gpu::ShuffleMode::XOR);
```

Comment:
Here we assume that all lanes are active? Is this always the case or can it happen that we don't evenly distribute across subgroups?

Reply:
We don't assume this. getShuffleOffset creates the strides in a way that inactive lanes are accounted for.